news/_posts/2019-09-11-prestashop-in-2019-and-beyond-part-2-pain-points.md

```diff
@@ -0,0 +1,471 @@
+
+Of course, after having invested time and money in making something work, developers won't appreciate having to update their code regularly every time there's a new platform version—ideally, it should "just work". Therefore, it goes without saying that introducing any change that breaks this agreed-upon _interface_ (a "breaking change") won't help with keeping developers interested in adopting a new version of said platform. Again, this is true for _all_ platforms.
+
+As a result, many platform vendors—PrestaShop included—will go out of their way to avoid introducing breaking changes as much as possible. But this is a double-edged sword.
+
+It is well-known that, in time, [software rots](https://en.wikipedia.org/wiki/Software_rot). Most developers agree that the best medicine for avoiding software decay is applying frequent [refactoring](https://refactoring.com/), because once code rots for good, it becomes unmaintainable and has to be rewritten—and history shows us that [rewriting from scratch is a terrible idea](https://www.joelonsoftware.com/2000/04/06/things-you-should-never-do-part-i/).
+
+However, refactoring can only get you so far until it starts to require introducing breaking changes. And once you hit that brick wall, you're stuck: either you let that part of the code rot or you introduce breaking changes that will undermine the adoption of your platform. Or worse, you introduce an _new_ subsystem that does the same thing as the old one, and keep the previous one for compatibility, consequently increasing the overall software complexity.
+
+That's one of the top reasons PrestaShop's architecture has become so complex. The constant need to provide as much backwards compatibility as possible, even when doing so was detrimental to the platform's overall technical quality, has substantially undermined the project's capacity to move forward.
+
+That said, let's find out what other problems are hiding under the hood...
+
+## Plan
+
+* [An aging architecture](#an-aging-architecture)
+* [No clearly-defined API](#no-clearly-defined-api)
+* [1.7 was released too soon](#17-was-released-too-soon)
+* [Too many dependencies with external services](#too-many-dependencies-with external-services)
+* [Testing is very hard](#testing-is-very-hard)
+* [Other problems](#other-problems)
+
+## An aging architecture
+
+PrestaShop has deeply-rooted design issues that can be explained simply by the age of its architecture.
+
+Web technologies have evolved massively in the last years, and industry standards are now very different compared to what they were 10 years ago, when the current foundations of PrestaShop were laid out. Back then, web applications were still being built using the classic backend-generated, form-centric approach that became commonplace during the [CGI-based era](https://en.wikipedia.org/wiki/Common_Gateway_Interface), which forged the modern Internet.
```

Comment:
Also PrestaShop has been built following the 10-years-old trend of "do your own MVC framework". All php pure players I have seen who were founded in this time did the same, because there was no standard framework popular enough* for people to use. 

*I know there were popular frameworks, but not wildly popular as Symfony and Laravel are now.